Add expression based test filtering

Guiguiprim opened the pull request:
As you are up apparently, I push this to open the discussion.

Following what you said, I was thinking maybe something along the line of:


DSL
 - `package(name)`
 - `deps(package_name)`
 - `rdeps(package_name)`
 - `test(partial_name)`
 - `not(expr)`
 - `expr && expr` / `expr and expr`
 - `expr || expr` / `expr or expr`
 - `(expr)`

Like `-E "package(driver) or test(use_driver_)"`, `-E "package(utility_package) and rdeps(utility_package)"`

sunshowers commented:
Your general design looks great. Let's do:
* `not x` to match `and` and `or`, along with a short form `! x`
* `x & y` as the short form for `x and y` (not `&&`) because this is similar to [`BitAnd` for sets](https://doc.rust-lang.org/alloc/collections/btree_set/struct.BTreeSet.html#impl-BitAnd%3C%26%27_%20BTreeSet%3CT%3E%3E)
* `x | y` and `x + y` for `x or y` for similar reasons
* `x - y` for `x and not y`
* `all()` for all tests

> `-E 'package(utility_package) and rdeps(utility_package)'`

Note that this performs the *intersection* of the two sets, which using the `and` operator honestly makes a bit less clear.

Some other questions:
* What sort of parser were you planning to use? My personal suggestion would be to use a parser combinator, and if you do so, [nom](https://docs.rs/nom/latest/nom/) would be my first suggestion (especially since nextest already uses it as a transitive dependency).
* How are you planning to do error handling? Ideally we'd integrate with something like [miette](https://docs.rs/miette/latest/miette/), following [this suggestion](https://www.reddit.com/r/rust/comments/p9t3jh/miette_a_fancy_new_diagnostics_definition_and/ha1mmyr/) by Kat.

Guiguiprim commented:
You are right, I meant all the tests in package `utility_package` and all the tests using `utility_package`, we should be careful not to make it error prone.

- I was thinking of `nom` !
- I didn't though about precedence yet. But I will be careful about it.
- I can look at `miette` 
- I had the same reflection about regexes, it adds parsing challenges, but yes `r'..'` looks right. I guess I would consider `literal:` default and optional in that case.
- Multiple expression sound fine.

"Sorting tests by key" you meant filtering on test names ? I guess I can keep it for the end, but as it might be useful for me I'm inclined to do it.

I probably won't have much time in the days to come, so this will be somehow slow to come.

sunshowers commented:
> "Sorting tests by key" you meant filtering on test names ? I guess I can keep it for the end, but as it might be useful for me I'm inclined to do it.

I mean sorting the order in which tests are run.

Guiguiprim commented:
I finally changed a bit the syntax:

- `package(name)` => contains `name`
- `package(=name)` => equal `name`
- `package(/name/)` => matches regex `name`

This makes it easier to parse and (I think) easier to use. `/../` is already a convention in some languages for regex.

The current limitations are:
- `equal` and `contains` can not contains `)`
- `regex` can not contains `/`

In both cases, this should not matter as they are both characters that can not be part of a test or package name.

Guiguiprim commented:
I have a first draft of filtering.

`deps()` and `rdeps()` are not implemented yet, but everything else is, so the following works:

- `-E "package(runner)"`
- `-E "package(=cargo-nextest)"`
- `-E "package(/nextest-.*/)"`
- `-E "test(parse) + test(basic) - test(env)"`
- `-E "all()"`
- `-E "none()"`
- `-E "test(parse) & test(expr_)"`
- `-E "test(expr_) | test(cargo)"`

Still todo:
- [ ] `deps()` / `rdeps()`
- [ ] Proper parsing error handling
- [ ] Support multiple `-E <expr>`
- [ ] Unit tests around matching

As of now, I have issues implementing `deps()`, [guppy::graph::DependsCache::depends_on()](https://docs.rs/guppy/latest/guppy/graph/struct.DependsCache.html#method.depends_on) doesn't seem to work...

Guiguiprim commented:
So it looks like I need to find out why my test of `-E "deps(runner)"` does not run any tests...

sunshowers commented on nextest-runner/src/test_filter/expression.rs:
```diff
@@ -0,0 +1,128 @@
+
+    /// Returns true if `artifact` is a dependency of at least
+    /// one workspace package matching `matcher`.
+    fn matches_dependency_of(artifact: &RustTestArtifact<'_>, matcher: &NameMatcher) -> bool {
+        let graph = artifact.package.graph();
+        let mut packages = graph
+            .packages()
+            .filter(|p| p.in_workspace())
+            .filter(|p| matcher.is_match(p.name()));
+
+        let id = artifact.package.id();
+        let mut cache = graph.new_depends_cache();
+
+        packages.any(|p| p.id() != id && cache.depends_on(p.id(), id).unwrap_or(false))
+    }
+}
```
This seems like it'll blow up the time complexity of the operation quite significantly:
* the number of artifacts is `O(number of workspace packages)`
* the complexity of graph traversal is `O(number of workspace packages + number of edges)`
* filtering by package is itself `O(number of workspace packages)`.

So it's worse than cubic in the number of workspace packages. For workspace-wide operations it's generally good to avoid anything past `O(n log n)` where `n` is the number of workspace packages.

Let's skip `deps` and `rdeps` operations for now and figure out how to implement them efficiently in a separate PR -- I think there's a bunch of ways we can rewrite this to make it more efficient.

Guiguiprim replied:
You're right, I didn't though about that.

And we (I) need to remember to remove `--no-deps` form our call to `cargo metadata`...

Guiguiprim replied:
We might "compile" the parsed expression using the packages graph to replace `deps()` and `rdeps()` by a set of accepted package id. This would reduce the cost inside `includes(artifact, test_name)`

Guiguiprim replied:
Rework with the compile approach. Let me known If you think this is better.

sunshowers commented on nextest-runner/src/test_filter/expression_parsing.rs:
```diff
@@ -0,0 +1,316 @@
+use super::expression::*;
```
Could you make sure to add license headers at the top of each file? They're important for code provenance.

Guiguiprim replied:
Done

sunshowers reviewed (commented):
This is looking great -- I'm really excited about how this is progressing!

Guiguiprim commented on nextest-runner/src/test_filter/expression.rs:
```diff
@@ -0,0 +1,128 @@
+    All,
+    /// No tests
+    None,
+    // Possible addition: Binary(NameMatcher)
```
That was the idea. And I wasn't going to do it in this iteration.

Guiguiprim commented:
I'm done with my compilation optimization pass, if you want to have another look.

sunshowers commented:
Thank you—I have other stuff coming up within the next couple of days but will try and take a look soon.

Guiguiprim commented:
I worked on parsing error with `miette`.

- it does make the parsing code more complex but the output is somewhat nice.
- I also moved all the filtering stuff in a new crate.
- I added some parsing tests

![parsing_error](https://user-images.githubusercontent.com/5973592/161109434-0b177a98-ebe6-4898-9641-7165fd20cb35.png)

I still need to add tests on the matching itself.

Guiguiprim commented:
Here we go, I:
- Added support for multiple expressions
- Added filtering tests
- Conditionally included `--no-deps` in the call to `cargo metadata`
- Made the feature experimental
- Added a first draft of documentation (this is not my stronger suit)

At this point all is working, and I'm ready for reviews and critics !

sunshowers reviewed (commented):
Thank you! It might take a little while for me to get through this, have some errands today and medical things for the rest of the week. The other maintainer is busy as well :)

Guiguiprim commented:
OK, noted, no problem.
Take care of yourself.

sunshowers commented on fixtures/tests-workspace-metadata.json:
```diff
@@ -0,0 +1 @@
+{"packages":[{"name":"crate_a","version":"0.1.0","id":"crate_a 0.1.0 (path+file:///home/fakeuser/tests-workspace/crate-a)","license":null,"license_file":null,"description":null,"source":null,"dependencies":[],"targets":[{"kind":["lib"],"crate_types":["lib"],"name":"crate_a","src_path":"/home/fakeuser/tests-workspace/crate-a/src/lib.rs","edition":"2021","doc":true,"doctest":true,"test":true}],"fea ...
```
Could you reformat this as pretty JSON?

sunshowers commented on nextest-filtering/src/parsing.rs:
```diff
@@ -0,0 +1,668 @@
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub(crate) enum Expr {
+    Not(Box<Expr>),
+    Union(Box<Expr>, Box<Expr>),
+    Intersection(Box<Expr>, Box<Expr>),
+    Set(SetDef),
+
+    Error,
+}
```
You could make this:

```
pub(crate) enum Expr {
    Valid(ValidExpr),
    Error,
}
```

and have `ValidExpr` be the valid expressions.

sunshowers commented on fixtures/tests-workspace/Cargo.toml:
```diff
@@ -0,0 +1,11 @@
+[workspace]
+members = [
+  "crate-a",
+  "crate-b",
+  "crate-c",
+  "crate-d",
+  "crate-e",
+  "crate-f",
+  "crate-g",
+]
+
```
Do we need this workspace checked into the repo?

sunshowers replied:
What about checking in a tarball?

sunshowers reviewed (approved):
Wow, this is absolutely incredible work. Thank you so much for doing it.

I think I'm going to try it out for a little bit locally and then land it. I don't really see any issues with it at first glance.

sunshowers commented on nextest-filtering/src/lib.rs:
```diff
@@ -0,0 +1,271 @@
+pub mod error;
+mod parsing;
+
+use parsing::{Expr, RawNameMatcher, SetDef};
+
+/// Matcher for name
+///
+/// Used both for package name and test name
+#[derive(Debug, Clone)]
+pub enum NameMatcher {
```
Could you move the contents of lib.rs into another file? The reason is that anything that's private here is fully accessible to any other modules.

sunshowers commented on nextest-filtering/src/lib.rs:
```diff
@@ -0,0 +1,271 @@
+/// Used to filter tests to run.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum FilteringExpr {
+    /// Accepts every tests not in the given expression
```
```suggestion
    /// Accepts every test not in the given expression
```

sunshowers commented on nextest-filtering/src/lib.rs:
```diff
@@ -0,0 +1,271 @@
+pub enum FilteringExpr {
+    /// Accepts every tests not in the given expression
+    Not(Box<FilteringExpr>),
+    /// Accepts every tests in either given expression
```
```suggestion
    /// Accepts every test in either given expression
```

sunshowers commented on nextest-filtering/src/lib.rs:
```diff
@@ -0,0 +1,271 @@
+    Not(Box<FilteringExpr>),
+    /// Accepts every tests in either given expression
+    Union(Box<FilteringExpr>, Box<FilteringExpr>),
+    /// Accepts every tests in both given expression
```
```suggestion
    /// Accepts every test in both given expressions
```

sunshowers commented on nextest-filtering/src/lib.rs:
```diff
@@ -0,0 +1,271 @@
+    Union(Box<FilteringExpr>, Box<FilteringExpr>),
+    /// Accepts every tests in both given expression
+    Intersection(Box<FilteringExpr>, Box<FilteringExpr>),
+    /// Accepts every tests in a set
```
```suggestion
    /// Accepts every test in a set
```

sunshowers commented on nextest-filtering/src/lib.rs:
```diff
@@ -0,0 +1,271 @@
+                    return Err(error::FilteringExprParsingError(input.to_string()));
+                }
+
+                let in_workspace_packages: Vec<_> =
+                    graph.packages().filter(|p| p.in_workspace()).collect();
```
You can just do `graph.resolve_workspace().packages(DependencyDirection::Forward).collect()` which will avoid an iteration over the full package graph. Not a huge deal.

sunshowers commented on nextest-filtering/src/lib.rs:
```diff
@@ -0,0 +1,271 @@
+                match compile_expr(&expr, &in_workspace_packages, &mut cache) {
+                    Some(expr) => Ok(expr),
+                    None => {
+                        // should not happen
```
Could you elaborate on `should not happen`?

sunshowers commented on nextest-filtering/src/lib.rs:
```diff
@@ -0,0 +1,271 @@
+                }
+            }
+            Err(_) => {
+                // should not happen
```
as well as here?

sunshowers reviewed (commented):
Just a few minor things.

sunshowers commented on nextest-filtering/src/error.rs:
```diff
@@ -0,0 +1,76 @@
+    }
+}
+
+#[derive(Debug)]
+pub struct FilteringExprParsingError(pub String);
```
Is there a way to make this an `ExpectedError` in the cargo-nextest crate? That will make it be printed as a nicer error (since this is a user error and not an internal error within nextest).

sunshowers commented on cargo-nextest/src/dispatch.rs:
```diff
@@ -308,8 +318,17 @@ impl TestBuildFilter {
         )?;
-        let test_filter =
-            TestFilterBuilder::new(self.run_ignored, self.partition.clone(), &self.filter);
+        let exprs = self
+            .expr_filter
+            .iter()
+            .map(|input| FilteringExpr::parse(input, graph))
+            .collect::<Result<Vec<_>, _>>()?;
```
Could this be reordered to happen before the build happens? This will cause parsing errors to be printed out immediately rather than after the build.

sunshowers reviewed (commented):
Using it locally now, a couple of initial comments.

sunshowers commented on nextest-filtering/src/parsing.rs:
```diff
@@ -0,0 +1,668 @@
+fn is_identifier_char(c: char) -> bool {
+    // This is use for NameMatcher::Contains(_) and NameMatcher::Equal(_)
+    // The output should be valid part of a test-name or a package name.
+    c == ':' || c.is_xid_continue()
```
Hmm, unfortunately this doesn't seem quite correct -- package names (and really test names with custom test harnesses) can have `-` in them, which is not XID_CONTINUE. In fact, test names can even be things like `foo.rs (line 30)` -- this is what doctests look like, for example.

I think we may want to accept arbitrary Unicode here for tests, and add ways to escape `(` and other special characters.

sunshowers commented on nextest-filtering/src/parsing.rs:
```diff
@@ -0,0 +1,668 @@
+}
+
+#[tracable_parser]
+fn parse_identifier_part(input: Span) -> IResult<Option<String>> {
```
Could you return a separate `Identifier` type? The constructor for this should accept a string and normalize it internally to NFC [the way Rust does](https://github.com/rust-lang/rfcs/blob/master/text/2457-non-ascii-idents.md#reference-level-explanation).

sunshowers commented on nextest-filtering/src/lib.rs:
```diff
@@ -0,0 +1,271 @@
+    for id1 in packages {
+        for p2 in all_packages {
+            let id2 = p2.id();
+            if id1 == id2 {
+                continue;
+            }
```
I'd match the behavior of https://bazel.build/reference/query here and in `rdependencies_packages` below and remove this check. `deps()` and `rdeps()` should return the transitive closure.

For a stretch goal, you could implement the depth limit listed there.

sunshowers commented on cargo-nextest/src/errors.rs:
```diff
@@ -131,6 +139,13 @@ impl ExpectedError {
                 );
                 None
             }
+            Self::FilterExpressionParseError { expressions } => {
+                log::error!(
+                    "Failed to parse some filtering expressions: {:?}",
+                    expressions
+                );
```
```suggestion
            Self::FilterExpressionParseError { .. } => {
                log::error!("failed to parse filter expression");
```

sunshowers commented on nextest-filtering/Cargo.toml:
```diff
@@ -0,0 +1,26 @@
+
+[features]
+default = []
+# Must be called `trace` which is a bit infortunate
```
```suggestion
# Must be called `trace` which is a bit unfortunate
```

sunshowers commented on site/src/book/filtering-expression.md:
```diff
@@ -0,0 +1,47 @@
+- can contains unicode sequence: `\u{xxx}` (where `xxx` is an 1 to 6 digits hexadecimal number)
+
+Operations:
+- `set_1 & set_2` , `set_1 and set_2`: include everything in both `set_1` and `set_2`
+- `set_1 | set_2`, `set_1 or set_2`, `set_1 + set_2`: include everything in either `set_1` or `set_2`
```
I'd be clear here and use the words "union" and "intersection".

sunshowers commented on site/src/book/filtering-expression.md:
```diff
@@ -0,0 +1,47 @@
+Operations:
+- `set_1 & set_2` , `set_1 and set_2`: include everything in both `set_1` and `set_2`
+- `set_1 | set_2`, `set_1 or set_2`, `set_1 + set_2`: include everything in either `set_1` or `set_2`
+- `not set`: include everything not included in `set`
```
`! set` works too I think!

sunshowers reviewed (approved):
Looking fantastic! Just a couple more changes and I'm happy to finish up the rest and get this out there!

sunshowers commented on site/src/book/filtering-expression.md:
```diff
@@ -0,0 +1,47 @@
+A filtering expression define a set of tests, any test in the set will be run.
+
+Basic sets:
+- `all()`: include everything
+- `test(name-matcher)`: include all tests matching `name-matcher`
+- `package(name-matcher)`: include all tests in packages matching `name-matcher`
+- `deps(name-matcher)`: include all tests in packages depended by packages matching `name-matcher` and include all tests in packages matching `name-matcher`
+- `rdeps(name-matcher)`: include all tests in packages depending on packages matching `name-matcher` and include all tests in packages matching `name-matcher`
+- `none()`: include nothing
```
Could you add a `partition` operator as well?

Guiguiprim replied:
Like `--partition`, this might be a little bit more complex and requires state.

Can we do it after, in an other PR ?

sunshowers commented on site/src/book/filtering-expression.md:
```diff
@@ -0,0 +1,47 @@
+- can contains unicode sequence: `\u{xxx}` (where `xxx` is an 1 to 6 digits hexadecimal number)
+
+Operations:
+- `set_1 & set_2` , `set_1 and set_2`: include everything in both `set_1` and `set_2`
+- `set_1 | set_2`, `set_1 or set_2`, `set_1 + set_2`: include everything in either `set_1` or `set_2`
+- `not set`: include everything not included in `set`
+- `set_1 - set_2`: equivalent to `set_1 and not set_2`
```
What is the operator precedence here -- are there tests for that? Could you document this?

Guiguiprim replied:
Doc added.

There is a test for it `nextest-filtering parsing::test_parse_epxr_precedence`

sunshowers commented on site/src/book/filtering-expression.md:
```diff
@@ -0,0 +1,47 @@
+
+Name matcher:
+- `text`: match anything containing `text`
+- `'unicode_text'`: match anything containing `unicode_text`
+- `=text`: match anything equal to `text`
+- `='unicode_text'`: match anything equal to `unicode_text`
```
Hmm, I don't think the single-quote scheme will work because on Unix shells the escaping becomes really awkward. I think we have to parse all text without using quote marks, then read input character by character, looking for the escapes you've listed below.

sunshowers commented on site/src/book/filtering-expression.md:
```diff
@@ -0,0 +1,52 @@
+- `/reg/`: match anything matching the regex `reg`
+
+Unicode string:
+- can contains escaped closing parenthesis: `\)`
+- can contains unicode sequence: `\u{xxx}` (where `xxx` is an 1 to 6 digits hexadecimal number)
```
Also `\r`, `\n` etc, right?

sunshowers commented on site/src/book/filtering-expression.md:
```diff
@@ -0,0 +1,52 @@
+
+Precedences from lowest to highest:
+- `not`, `()`
+- `&`, `|`, `-`
```
How does an expression with `&` and `|` get resolved? In general I'd expect us to follow a similar pattern to https://doc.rust-lang.org/reference/expressions.html#expression-precedence -- from highest to lowest:

* `()`
* `!`, `not`
* `&`, `and`
* `|`, `+`, `or`

Only real difference is that `+` is grouped with `|`.

sunshowers reviewed (approved):
Looks great -- just a couple more changes, and I'm happy to land this and finish it up myself if you like.

sunshowers commented:
Going to land this and clean it up, thanks again for all your incredible efforts here.
